File: 42_topological_sort/42_4_counting_paths.cpp
#include <iostream>
#include <vector>
#include <map>

std::vector<int> topological_sort(const std::vector<std::vector<int>>& graph) {
  int v = graph.size();

  std::vector<int> in_degrees(v, 0);
  for (int node = 0; node < v; node++) {
    for (int nbr : graph[node]) {
      in_degrees[nbr]++;
    }
  }

  std::vector<int> zero_degree;
  for (int node = 0; node < v; node++) {
    if (in_degrees[node] == 0) {
      zero_degree.push_back(node);
    }
  }

  std::vector<int> topo_order;
  while (!zero_degree.empty()) {
    int node = zero_degree.back();
    zero_degree.pop_back();
    topo_order.push_back(node);
    
    for (int nbr : graph[node]) {
      in_degrees[nbr]--;
      if (in_degrees[nbr] == 0) {
        zero_degree.push_back(nbr);
      }
    }
  }

  if (topo_order.size() < v) {
    return std::vector<int>();
  }

  return topo_order;
}

std::vector<int> counting_paths(const std::vector<std::vector<int>>& graph, int start) {
  std::vector<int> topo_order = topological_sort(graph);
  std::map<int, int> counts;
  counts[start] = 1;

  for (int node : topo_order) {
    if (counts.find(node) == counts.end()) continue;

    for (int nbr : graph[node]) {
      if (counts.find(nbr) == counts.end()) {
        counts[nbr] = counts[node];
      } else {
        counts[nbr] += counts[node];
      }
    }
  }

  std::vector<int> res;
  for (int node = 0; node < graph.size(); node++) {
    if (counts.find(node) == counts.end()) {
      res.push_back(0);
    } else {
      res.push_back(counts[node]);
    }
  }

  return res;
}

int main() {
    // Example
    std::vector<std::vector<int>> graph = {
        {1},
        {},
        {1},
        {4},
        {1, 2, 5},
        {2}
    };
    int start = 4;

    std::vector<int> result = counting_paths(graph, start);

    for (int i = 0; i < result.size(); i++) {
        std::cout << result[i];
        if (i < result.size() - 1) {
            std::cout << " ";
        }
    }
    std::cout << std::endl;

    return 0;
}

File: 42_topological_sort/42_4_counting_paths.py
def topological_sort(graph: list[list[int]])->list[int]:
  v = len(graph)
  in_degrees = [0] * v
  for node in range(v):
    for nbr in graph[node]:
      in_degrees[nbr] += 1

  zero_degree = []
  for node in range(v):
    if in_degrees[node] == 0:
      zero_degree.append(node)

  topo_order = []
  while len(zero_degree) > 0:
    node = zero_degree.pop()
    topo_order.append(node)
    for nbr in graph[node]:
      in_degrees[nbr] -= 1
      if in_degrees[nbr] == 0:
        zero_degree.append(nbr)
    
  if len(topo_order) < v:
    return []

  return topo_order

def counting_paths(graph: list[list[int]], start: int)->list[int]:
  topo_order = topological_sort(graph)
  counts = {start: 1}
  for node in topo_order:
    if node not in counts: continue
    for nbr in graph[node]:
      if nbr not in counts:
        counts[nbr] = counts[node]
      else:
        counts[nbr] += counts[node]
  
  res = []

  for node in range(len(graph)):
    if node not in counts:
      res.append(0)
    else:
      res.append(counts[node])

  return res

if __name__== "__main__":
  # Example
  graph = [
    [1],
    [],
    [1],
    [4],
    [1, 2, 5],
    [2]
  ]
  start = 4
  print(counting_paths(graph, start))